File: benchmarks/binary_trees/binary_trees.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Tree {
    struct Tree *left;
    struct Tree *right;
} Tree;

static Tree *make(int depth) {
    Tree *t = (Tree *)malloc(sizeof(Tree));
    if (depth == 0) {
        t->left = NULL;
        t->right = NULL;
    } else {
        t->left  = make(depth - 1);
        t->right = make(depth - 1);
    }
    return t;
}

static long long check(Tree *t) {
    if (t == NULL) return 0;
    return 1 + check(t->left) + check(t->right);
}

static void free_tree(Tree *t) {
    if (t == NULL) return;
    free_tree(t->left);
    free_tree(t->right);
    free(t);
}

int main(void) {
    int min_depth = 4;
    int max_depth = 14;
    long long total = 0;

    /* stretch tree */
    int stretch_depth = max_depth + 1;
    Tree *stretch = make(stretch_depth);
    total += check(stretch);
    free_tree(stretch);

    /* long-lived tree */
    Tree *long_lived = make(max_depth);

    /* iteration loop */
    for (int depth = min_depth; depth <= max_depth; depth += 2) {
        long long exp = max_depth - depth + min_depth;
        long long iterations = 1LL << exp;
        for (long long i = 0; i < iterations; i++) {
            Tree *t = make(depth);
            total += check(t);
            free_tree(t);
        }
    }

    /* long-lived tree check */
    total += check(long_lived);
    free_tree(long_lived);

    printf("%lld\n", total);
    return 0;
}

File: benchmarks/binary_trees/binary_trees.rs
// binary_trees — allocation-churn benchmark (mirrors binary_trees.c / .rl).
// maxDepth=14; builds, checks and drops many short-lived trees plus one
// long-lived tree. Box<Tree> gives malloc/free-style ownership (RAII drop),
// the Rust analogue of C's malloc/free and Rolang's ARC.

struct Tree {
    left: Option<Box<Tree>>,
    right: Option<Box<Tree>>,
}

fn make(depth: i32) -> Box<Tree> {
    if depth == 0 {
        Box::new(Tree { left: None, right: None })
    } else {
        Box::new(Tree {
            left: Some(make(depth - 1)),
            right: Some(make(depth - 1)),
        })
    }
}

fn check(t: &Tree) -> i64 {
    let mut sum: i64 = 1;
    if let Some(ref l) = t.left {
        sum += check(l);
    }
    if let Some(ref r) = t.right {
        sum += check(r);
    }
    sum
}

fn main() {
    let min_depth: i32 = 4;
    let max_depth: i32 = 14;
    let mut total: i64 = 0;

    // stretch tree
    let stretch_depth = max_depth + 1;
    let stretch = make(stretch_depth);
    total += check(&stretch);
    drop(stretch);

    // long-lived tree
    let long_lived = make(max_depth);

    // iteration loop
    let mut depth = min_depth;
    while depth <= max_depth {
        let exp = max_depth - depth + min_depth;
        let iterations: i64 = 1i64 << exp;
        for _ in 0..iterations {
            let t = make(depth);
            total += check(&t);
        }
        depth += 2;
    }

    // long-lived tree check
    total += check(&long_lived);

    println!("{}", total);
}
